### ISSUE TITLE
TIOCGWINSZ and CSI 14 may return different geometries at startup, and until a resize event

### ISSUE THREAD

**Issue by csdvrx** [mintty/mintty#1071]
# Description:

On the latest msys2, the geometry returned by TIOCGWINSZ  doesn't always match CSI 14.

After maximizing mintty by doubleclicking on the titlebar then double clicking again to return to normal,  the geometry becomes consistent.

# Setup:

For example, starting at 88x27 using Iosevka 18pt on a 1800x1200 screen with windows scale and layout at 150%, with no bash prompt to avoid confusion, then using a measuring script before each dimensioning, it becomes obvious the first size is wrong in TIOCGWINSZ.

The size is corrected when the window is demaximized, which resynchronizes the results of TIOCGWINSZ with CSI 14

# Tools

Script:
```
#!/bin/bash
STATE=$1
# check the length matches
printf "%0.s-" {1..88}
echo
# compare the results
echo -n "now $STATE, TIOCGWINSZ: "
/d/wingetsize.exe
echo -n "OSC 14: "
echo -ne "\\033[14t\\033\\ "
```

Sourcecode of wingetsize.c:

```
#include <sys/ioctl.h>
#include <stdio.h>

int main (void)
{
    struct winsize ws;
    ioctl(1, TIOCGWINSZ, &ws);
    printf("rows: %d cols: %d xpixel: %d ypixel: %d\n", ws.ws_row, ws.ws_col, ws.ws_xpixel, ws.ws_ypixel);
}
```

# Test protocol:

 - step 0: start with no bash prompt to avoid confusion: unset PS0 PS1 PROMPT_COMMAND ; stty intr ^X ; reset
 
 - step 1: query the geometry

 - step 2: maximize

 - step 3: return to normal

# Evidence:

Matching screenshot of the full screen (to facilitate visual comparisons) for each step: notice how xpixel, ypixel and OSC14 are initially off:

<img width="900" alt="step0" src="https://user-images.githubusercontent.com/49044530/102289675-db80c300-3f04-11eb-8a9c-d25afa7b02f7.png">

<img width="900" alt="step1" src="https://user-images.githubusercontent.com/49044530/102289683-e0de0d80-3f04-11eb-8ae2-d62c85e35029.png">

<img width="900" alt="step2" src="https://user-images.githubusercontent.com/49044530/102289689-e3406780-3f04-11eb-99b1-eb5bc0141325.png">

<img width="900" alt="step3" src="https://user-images.githubusercontent.com/49044530/102289695-e76c8500-3f04-11eb-8a1d-cb10db8fe834.png">

Reducing the initial number of lines doesn't help:

<img width="289" alt="menu" src="https://user-images.githubusercontent.com/49044530/102290119-f3a51200-3f05-11eb-9dae-da08f3eed524.png">

As can be seen, the issue persists with one less line:

<img width="900" alt="step0-oneless" src="https://user-images.githubusercontent.com/49044530/102290146-061f4b80-3f06-11eb-8a69-e1e0f6bf84bf.png">

The issue persists even with a very different geometry and the taskbar set to normal:

<img width="900" alt="stop0-with-taskbar" src="https://user-images.githubusercontent.com/49044530/102290687-23a0e500-3f07-11eb-907e-b2012104af1b.png">

As can be seen in this last example, it may extend to xpixel as well.

# Workaround:

Any change of dimensions (maximizing, but also increasing the canvas by just one line) resynchronizes both results.

**Comment by csdvrx** [comment]
Apparently it's called CSI so I'm correcting it

**Comment by mintty** [comment]
Using the fields ws_ypixel and ws_xpixel is unreliable, e.g. the Linux manual says that they are unused (https://www.man7.org/linux/man-pages/man4/tty_ioctl.4.html).
Nonetheless, it is inconsistent that mintty fills them with wrong initial values.

**Comment by avih** [comment]
In my experience (overseeing adding sixel support to mpv), we concluded that all OSs we had access to support these fields, and when the terminal reports them correctly, then they're correct for reading by other applications (using an application similar to the one listed earlier here).

We looked at linux, free/net/open bsd, illumos, osx, cygwin  - in all of them these fields are usable.

We also looked at the terminals xterm, mlterm, tmux, st, alacritty, and others, and except for konsole which doesn't report pixel values, and VTE which only reports them when built with sixel support, all others report the pixel values.

**Comment by mintty** [comment]
Fixed in repository.

**Comment by BrianInglis** [comment]
> In my experience (overseeing adding sixel support to mpv), we concluded that all OSs we had access to support ws_xpixel/ws_ypixel, and when the terminal reports them correctly, then they're correct for reading by other applications (using an application which uses `ioctl` similar to the one listed earlier here).

What pixel values are correct in this case? Same as CSI 14 t?

> We looked at linux, free/net/open bsd, illumos, osx, cygwin - in all of them these fields are usable.
>
> We also looked at the terminals xterm, mlterm, tmux, st, alacritty, iTerm2, and others, and except for konsole which doesn't report pixel values, and VTE which only reports them when built with sixel support, all others report the pixel values (including panes in tmux, as long as the containing outer-terminal reports them).

Some articles said that pixel return values differed, often including character and window padding, sometimes window decorations including scroll bars, equivalent to values reported by other variants of CSI # t.

**Comment by mintty** [comment]
To deal with this issue, xterm has introduced (not long ago) a variation described as follows:
>            Ps = 1 4  -> Report xterm text area size in pixels.
>            Ps = 1 4 ;  2  -> Report xterm window size in pixels.
> Normally xterm's window is larger than its text area, since it includes the frame (or decoration) applied by the window manager, as well as the area used by a scroll-bar.

So I think, following xterm as an archetype of terminal emulators, CSI 14 t should report the raw character cell space without any padding and decoration.

For the ioctl, specification is fuzzy, but concluding from the fact that the pixel values are bundled with the row and column counts in one struct, the same should apply.

**Comment by avih** [comment]
> What pixel values are correct in this case? Same as CSI 14 t?

Whatever the terminal sets via ioctl - i.e. the pixel members are useful and usable for an application to read the size which the terminal reports. I didn't check CSI 14.

> Some articles said that pixel return values differed, often including character and window padding, sometimes window decorations including scroll bars, equivalent to values reported by other variants of CSI # t.

I haven't tested CSI # t, but indeed the values are not consistent. mlterm and st report the area size covered by the cells (excluding padding and anything else), xterm includes the padding outside the cells, I believe VTE also reports the area covered by the cells, and additionally, some terminals (xterm/mlterm) report slightly different when maximized or fullscreen.

I didn't test all terminals if the padding is included, in part because the heuristics we ended up deploying works reasonably well (rounding the pixel value to a multiple of the number of cells for each axis worked well enough - but it does have limits). We were only interested in the area usable with sixel - which is the area covered by the cells.

**Comment by avih** [comment]
FWIW, IMO the reported size in pixels should be the one covered by the cells. Not because sixels needs it (and it does need it), but because the area outside the cells is not inherently content which a terminal displays. It's decoration etc, and can vary between terminals, window managers, etc, configuration values, etc.

**Comment by BrianInglis** [comment]
When I wrote CSI *#* t I meant *#* in the US sense of numeric variable e.g. one of my bash functions reporting terminal info including CSI *11...19* t:
```
$ xterm_info
type features   VT420wpSeNThcrl ^[[c            ^[[?64;1;2;4;6;9;15;21;22;28;29c$
type version    mintty v3.4.3-0 ^[[>c           ^[[>77;30403;0c$
window position 3 px 210+120    ^[[13t          ^[[3;210;120t$
window size     4 px 875x881    ^[[14;2t        ^[[4;881;875t$
text position   3 px 219+152    ^[[13;2t        ^[[3;219;152t$
text size       4 px 840x840    ^[[14t          ^[[4;840;840t$
text chars      8 ch 120x60     ^[[18t          ^[[8;60;120t$
char size       6 px 7x14       ^[[16t          ^[[6;14;7t$
screen chars    9 ch 274x85     ^[[19t          ^[[9;85;274t$
screen size     5 px 1920x1200  ^[[15t          ^[[5;1200;1920t$
window status   1 open          ^[[11t          ^[[1t$
```

**Comment by csdvrx** [comment]
> overseeing adding sixel support to mpv

@avih, thanks a lot for your great work BTW. I'm using mpv -> sixel -> tmux-sixel/derasterize hybrid to play videos in unicode 

> rounding down the pixel value to a multiple of the number of cells for each axis worked well enough

simple and practical. will borrow :)

**Comment by mintty** [comment]
Released 3.4.4.
